sample_project/Assets/SampleViewer/Samples/FeatureLayer/Scripts/FeatureLayerInputManager.cs

```diff
@@ -0,0 +1,200 @@
+using TMPro;
+using UnityEngine;
+using UnityEngine.InputSystem;
+
+public class FeatureLayerInputManager : MonoBehaviour
+{
+    public FeatureData FeatureData;
+
+    [SerializeField] private Transform container;
+    private InputActions inputActions;
+    private bool isLeftShiftPressed;
+    private List<GameObject> items = new List<GameObject>();
+    [SerializeField] private Material outlineMat;
+    [SerializeField] private GameObject properties;
+    [SerializeField] private GameObject propertiesView;
+    private TouchControls touchControls;
+    private FeatureLayerUIManager uiManager;
+
+    private IEnumerator AddItemsToScrollView()
+    {
+        foreach (var item in items)
+        {
+            SetScrollViewItems(item);
+            yield return new WaitForSeconds(0.01f);
+        }
+
+        StopCoroutine("AddItemsToScrollView");
+    }
+
+    private void Awake()
+    {
+#if !UNITY_IOS && !UNITY_ANDROID && !UNITY_VISIONOS
+        inputActions = new InputActions();
+#else
+        touchControls = new TouchControls();
+#endif
+    }
+
+    public void ClearAdditionalMaterial(GameObject feature)
+    {
+        Material[] materialsArray = new Material[feature.GetComponentInChildren<Renderer>().materials.Length - 1];
+
+        for (int i = 0; i < feature.GetComponentInChildren<Renderer>().materials.Length - 1; i++)
+        {
+            materialsArray[i] = feature.GetComponentInChildren<Renderer>().materials[i];
+        }
+
+        feature.GetComponentInChildren<Renderer>().materials = materialsArray;
+    }
+
+    public void EmptyPropertiesDropdown()
+    {
+        if (items == null)
+        {
+            return;
+        }
+
+        foreach (var item in items)
+        {
+            Destroy(item.gameObject);
+        }
+
+        items.Clear();
+    }
+
+    private void OnEnable()
+    {
+#if !UNITY_IOS && !UNITY_ANDROID && !UNITY_VISIONOS
+        inputActions.Enable();
+        inputActions.DrawingControls.LeftClick.started += OnLeftClickStart;
+        inputActions.DrawingControls.LeftShift.performed += ctx => OnLeftShift(true);
+        inputActions.DrawingControls.LeftShift.canceled += ctx => OnLeftShift(false);
+#else
+        touchControls.Enable();
+        touchControls.Touch.TouchPress.started += OnTouchInputStarted;
+#endif
+    }
+
+    private void OnDisable()
+    {
+#if !UNITY_IOS && !UNITY_ANDROID && !UNITY_VISIONOS
+        inputActions.Disable();
+        inputActions.DrawingControls.LeftClick.started -= OnLeftClickStart;
+        inputActions.DrawingControls.LeftShift.performed -= ctx => OnLeftShift(true);
+        inputActions.DrawingControls.LeftShift.canceled -= ctx => OnLeftShift(false);
+#else
+        touchControls.Disable();
+        touchControls.Touch.TouchPress.started -= OnTouchInputStarted;
+#endif
+    }
+
+    private void OnLeftShift(bool isPressed)
+    {
+        isLeftShiftPressed = isPressed;
+    }
+
+    private void OnLeftClickStart(InputAction.CallbackContext obj)
```

Comment:
This and `OnTouchImputStarted` are almost the same and can get refactored so you're not repeating yourself

Reply:
They can't. The input actions are different and have different bindings. This is how we do it in all the other samples in order to properly support touch input in samples.

Reply:
Right, Im not saying get rid of the funcs, but the internal code could be refactored something like this:

```
private void OnLeftClickStart(InputAction.CallbackContext obj)
    {
        if (!isLeftShiftPressed)
        {
            return;
        }

        Ray ray = Camera.main.ScreenPointToRay(Mouse.current.position.ReadValue());
        HandleInput(ray);
    }

    private void OnTouchInputStarted(InputAction.CallbackContext ctx)
    {
        Ray ray = Camera.main.ScreenPointToRay(touchControls.Touch.TouchPosition.ReadValue<Vector2>());
        HandleInput(ray);
    }

    private void HandleInput(Ray ray)
    {
        RaycastHit hit;
        if (Physics.Raycast(ray, out hit))
        {
            if (FeatureData)
            {
                ClearAdditionalMaterial(FeatureData.gameObject);
            }

            FindFirstObjectByType<FeatureLayerUIManager>().dropDownButton.isOn = false;
            EmptyPropertiesDropdown();
            FeatureData = hit.collider.gameObject.GetComponent<FeatureData>();

            if (!FeatureData)
            {
                return;
            }

            var featureLayer = FindFirstObjectByType<FeatureLayer>();
            featureLayer.RefreshProperties(FeatureData.gameObject);

            foreach (var property in FeatureData.Properties)
            {
                var item = Instantiate(properties);
                items.Add(item);
                item.GetComponentInChildren<TextMeshProUGUI>().text = property;
            }

            StartCoroutine("AddItemsToScrollView");
            SetAdditionalMaterial(outlineMat, hit.collider);
            propertiesView.SetActive(true);
        }
    }
```

Reply:
Gotcha, that makes sense.